proposals/0000-ghc-cpp.rst

```diff
@@ -0,0 +1,231 @@
+.. sectnum::
+.. contents::
+
+Haskell code uses the CPP pragma extensively. I'm proposing to
+introduce a new GhcCPP pragma to perform a subset of CPP processing as
+an alternative, as part of the normal GHC compilation process.
+
+This subset will be just enough to control conditional compilation,
+and will only be active in lines introduced with a ``#`` and
+preprocessor keyword.
+
+It will be tooling friendly, in that parts of the code disabled by the
+processor will be saved as comments, if comments are being processed.
+
+
+Motivation
+----------
+
+Haskell code uses the CPP pragma to manage version changes, when the
+compiler or a library changes.
+
+CPP gets the job done, but it is not haskell-friendly, and so has a
+side-effect of mangling otherwise normal haskell code, which is not
+intended to be processed for the version-selection process.
+
+It is also not tooling-friendly. It runs as a separate pass over the
+source code, emitting updated source for compilation which can
+actually change pretty much anything, if a person decides to use the
+full power of CPP to redefine existing Haskell, or introduce macros to
+generate boilerplate code.
+
+The primary motivation for this proposal is to make it possible to
+safely perform source-code transformations on code containing
+preprocessor directives.
+
+Currently in ``ghc-exactprint`` when a file uses CPP, the original
+file and preprocessor output are diffed at the token level and the
+"missing" parts inserted as comments. This works in some cases, but
+for many it fails, for reasons such as
+
+- A comparison to a numeric constant:
+
+.. code:: haskell
+
+   #if __GLASGOW_HASKELL__ < 707
+
+Shows up as a stray ``906`` if compiled with GHC 9.6.x
+
+- Preprocessor directives inside haskell comments are also processed.
+  This makes the diff fail too.
+
+- Trailing backslashes are eliminated, and the lines joined up. This
+  breaks some ASCII art comments, as well as multi-line strings.
+
+By implementing a subset of CPP processing inside GHC, we can avoid
+these drawbacks, and provide conditional compilation facilities that
+cover most of the "normal" current needs.
+
+The processing will be explicitly limited to lines introduced by a
+standard CPP directive, such as ``#define FOO``, and operate at the
+GHC Lexer level. This means that it will not touch directives
+appearing inside comments, nor will CPP macros be applies anywhere but
+in other CPP directives.
+
+
+Proposed Change Specification
+-----------------------------
+
+When enabled, the lexer has the additional rules:
+
+.. code:: haskell
+
+   cpp -> '#' cppkeyword [ anysymbol {any} ] newline
+         | '#' cppkeyword [ anysymbol {any} ] '\' newline cppcont
+   cppcont -> [ anysymbol {any} ] '\' newline cppcont
+         |  [ anysymbol {any} ] newline
+   cppkeyword -> 'define' | 'include' | 'undef' | 'error' | 'ifdef'
+                 | 'ifndef' | 'if' | 'elif' | 'else' | 'endif'
+
+   ccomment = '/*'  [ anysymbol {any} ] '*/'
```

Comment:
What's the point of supporting another type of comment?

Reply:
`#include` files (unfortunately for us) have C-style comments. E.g. `cabal_macros.h`, `ghcautoconf.h`